hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -198,111 +214,135 @@ public Object convert(Object value) {
-        }
+        public static final QueryDataTypeSerializer INSTANCE = new QueryDataTypeSerializer();
 
-        writeObjectTypeMetadata(this, out);
-        if (!isCustomType()) {
-            return;
+        @Override
+        public int getFactoryId() {
+            return SqlDataSerializerHook.F_ID;
         }
 
-        Map<String, QueryDataType> typeMap = new HashMap<>();
-        collectCustomTypes(this, typeMap);
-
-        out.writeInt(typeMap.size());
-        for (QueryDataType nestedType : typeMap.values()) {
-            writeObjectTypeMetadata(nestedType, out);
-            out.writeInt(nestedType.getObjectFields().size());
-            for (QueryDataTypeField field : nestedType.getObjectFields()) {
-                out.writeString(field.name);
-                out.writeInt(field.type.converter.getId());
-                writeObjectTypeMetadata(field.type, out);
-            }
+        @Override
+        public int getClassId() {
+            return SqlDataSerializerHook.QUERY_DATA_TYPE;
         }
-    }
 
-    private static void writeObjectTypeMetadata(QueryDataType type, ObjectDataOutput out) throws IOException {
-        out.writeInt(type.objectTypeKind.value());
-        out.writeString(type.objectTypeName);
-        out.writeString(type.objectTypeMetadata);
-    }
+        /**
+         * @implNote Collects all distinct custom types into a <em>type map</em> beforehand
+         * to avoid infinite recursion. Then, it writes each type with its direct children,
+         * i.e. each subtree, in an arbitrary order. {@link #read} first creates a type map
+         * that initially contains only this {@code QueryDataType}, i.e. the root. Then, it
+         * reads all subtrees by creating a type only if it is not created before using the
+         * type map. Even though subtrees don't lie in a particular order, the children of
+         * all subtrees will eventually be populated, including the root.
+         */
+        @Override
+        public void write(ObjectDataOutput out, QueryDataType type) throws IOException {
+            out.writeInt(type.converter.getId());
+            if (type.converter.getTypeFamily() != QueryDataTypeFamily.OBJECT) {
+                return;
+            }
 
-    private void collectCustomTypes(QueryDataType type, Map<String, QueryDataType> typeMap) {
-        typeMap.put(type.objectTypeName, type);
+            writeObjectTypeMetadata(type, out);
+            if (!type.isCustomType()) {
+                return;
+            }
 
-        for (QueryDataTypeField field : type.objectFields) {
-            if (field.getType().isCustomType()) {
-                if (!typeMap.containsKey(field.type.objectTypeName)) {
-                    collectCustomTypes(field.type, typeMap);
+            Map<String, QueryDataType> typeMap = new HashMap<>();
+            collectCustomTypes(type, typeMap);
+
+            out.writeInt(typeMap.size());
+            for (QueryDataType nestedType : typeMap.values()) {
```

Comment:
@burakgok can you point me, where exactly is it tested?

Reply:
https://github.com/hazelcast/hazelcast/blob/df6bbea57d878cf70c75fe129ef5308164694a94/hazelcast-sql/src/test/java/com/hazelcast/sql/impl/type/QueryDataTypeTest.java#L215-L228